File: src/ruby/lib/geodelta/packer32.rb
# encoding: utf-8

module GeoDelta
  class Packer32
    def pack_world_delta(id)
      return id << 28
    end

    def unpack_world_delta(value)
      return (value >> 28) & 0b111
    end

    def pack_sub_delta(level, id)
      return id << (26 - ((level - 2) * 2))
    end

    def unpack_sub_delta(level, value)
      return (value >> (26 - ((level - 2) * 2))) & 0b11
    end

    def pack_level(level)
      return level
    end

    def unpack_level(value)
      return value & 0b1111
    end

    def pack(ids)
      wid   = self.pack_world_delta(ids[0])
      sids  = ids[1..-1].each_with_index.map { |id, i| self.pack_sub_delta(i + 2, id) }.inject(0, &:+)
      level = self.pack_level(ids.size)
      return wid + sids + level
    end

    def unpack(value)
      level = self.unpack_level(value)
      wid   = self.unpack_world_delta(value)
      sids  = (level - 1).times.map { |i| self.unpack_sub_delta(i + 2, value) }
      return [wid] + sids
    end
  end
end

File: src/java/geodelta/src/jp/nayutaya/geodelta/Packer32.java

package jp.nayutaya.geodelta;

/**
 * 32ビット符号付き整数にパックされたデルタID列を扱うクラス。
 */
public class Packer32
{
    /**
     * ワールドデルタIDをパックする。
     *
     * @param id ワールドデルタID
     * @return パックされたID
     */
    /* package */static int packWorldDelta(final int id)
    {
        return id << 28;
    }

    /**
     * ワールドデルタIDをアンパックする。
     *
     * @param value パックされたデルタID列
     * @return ワールドデルタID
     */
    /* package */static byte unpackWorldDelta(final int value)
    {
        return (byte)((value >> 28) & 0x07);
    }

    /**
     * サブデルタIDをパックする。
     *
     * @param level サブデルタのレベル
     * @param id サブデルタID
     * @return パックされたID
     */
    /* package */static int packSubDelta(final int level, final int id)
    {
        return id << (26 - ((level - 2) * 2));
    }

    /**
     * サブデルタIDをアンパックする。
     *
     * @param level サブデルタのレベル
     * @param value パックされたデルタID列
     * @return サブデルタID
     */
    /* package */static byte unpackSubDelta(final int level, final int value)
    {
        return (byte)((value >> (26 - ((level - 2) * 2))) & 0x03);
    }

    /**
     * レベルをパックする。
     *
     * @param level レベル
     * @return パックされたID
     */
    /* package */static int packLevel(final int level)
    {
        return level;
    }

    /**
     * レベルをアンパックする。
     *
     * @param value パックされたデルタID列
     * @return レベル
     */
    /* package */static int unpackLevel(final int value)
    {
        return value & 0x0F;
    }

    /**
     * デルタID列を32ビット整数にパックする。
     *
     * @param ids デルタID列
     * @return パックされたデルタID列
     */
    public static int pack(final byte[] ids)
    {
        final int wid = packWorldDelta(ids[0]);
        int sid = 0;
        for ( int i = 1, len = ids.length; i < len; i++ )
        {
            sid |= packSubDelta(i + 1, ids[i]);
        }
        final int level = packLevel(ids.length);
        return wid | sid | level;
    }

    /**
     * 32ビット整数からデルタID列をアンパックする。
     *
     * @param value パックされたデルタID列
     * @return デルタID列
     */
    public static byte[] unpack(final int value)
    {
        final int level = unpackLevel(value);
        final byte[] ids = new byte[level];
        ids[0] = unpackWorldDelta(value);
        for ( int i = 1; i < level; i++ )
        {
            ids[i] = unpackSubDelta(i + 1, value);
        }
        return ids;
    }
}
